tests/unit/stream_alert_rule_processor/test_sink.py

```diff
@@ -96,17 +103,131 @@ def test_streamsink_sink_success(self, log_mock):
+        return [
+            {
+                'record': '{"abc": 123}',
+                'rule_description': 'Desc{}'.format(i),
+                'rule_name': 'Rule{}'.format(i),
+                'outputs': ['aws-lambda:...', 'aws-s3:...']
+            }
+            for i in xrange(count)
+        ]
+
+    def test_alert_batches_single_alert(self):
+        """StreamSink - Alert Batching - Single Alert"""
+        alerts = self._generate_alerts(1)
+        result = list(self.sinker._alert_batches(alerts))
+
+        # Replace timestamp to allow for equality checking.
+        result[0][self.ALERT_TABLE][0]['PutRequest']['Item']['Timestamp']['S'] = 'now'
+
+        expected = [
+            {
+                self.ALERT_TABLE: [
+                    {
+                        'PutRequest': {
+                            'Item': {
+                                'RuleName': {'S': 'Rule0'},
+                                'Timestamp': {'S': 'now'},
+                                'Cluster': {'S': 'corp'},
+                                'RuleDescription': {'S': 'Desc0'},
+                                'Outputs': {'SS': ['aws-lambda:...', 'aws-s3:...']},
+                                'Record': {'S': '"{\\"abc\\": 123}"'}
+                            }
+                        }
+                    }
+                ]
+            }
+        ]
+        assert_equal(expected, result)
+
+    def test_alert_batches_max_batch(self):
+        """StreamSink - Alert Batching - Full Batch"""
+        alerts = self._generate_alerts(10)
+        result = list(self.sinker._alert_batches(alerts, batch_size=10))
+
+        assert_equal(1, len(result))  # There should be one batch.
+        assert_equal(10, len(result[0][self.ALERT_TABLE]))  # All 10 alerts should be in the batch.
+
+    def test_alert_batches_max_batch_plus_one(self):
+        """StreamSink - Alert Batching - Full Batch + 1"""
+        alerts = self._generate_alerts(11)
+        result = list(self.sinker._alert_batches(alerts, batch_size=10))
+
+        assert_equal(2, len(result))  # There should be 2 alert batches.
+        assert_equal(10, len(result[0][self.ALERT_TABLE]))  # 10 alerts in the first batch.
+        assert_equal(1, len(result[1][self.ALERT_TABLE]))  # 1 alert in the second batch.
+
+    @patch('stream_alert.rule_processor.sink.time')
+    @patch('stream_alert.rule_processor.sink.LOGGER')
+    def test_dynamo_sink_unprocessed_alerts(self, log_mock, time_mock):
+        """StreamSink - Dynamo Sink - Retry unprocessed alerts"""
+
+        def mock_batch_write_item(**kwargs):
+            """Mock client_dynamo.batch_write_item to always return all items unprocessed."""
+            return {
+                'ResponseMetadata': {'HTTPStatusCode': 200},
+                'UnprocessedItems': {self.ALERT_TABLE: kwargs['RequestItems']}
+            }
+
+        self.sinker.client_dynamo.batch_write_item = mock_batch_write_item
```

Comment:
I'd suggest using `patch` for this if possible (which I think it should be). If not, ignore me.. :)

You should be able to do this like so:

```
@patch('boto3.client')  ## add your other patches also...
def test_dynamo_sink_unprocessed_alerts(boto_mock):
    """Use return_value to specify what you want a patched object to return"""
    alerts = self._generate_alerts(1)
    value = {
        'ResponseMetadata': {'HTTPStatusCode': 200},
        'UnprocessedItems': {self.ALERT_TABLE: alerts}
    }
    
    boto_mock.return_value.batch_write_item.return_value = value

    self.sinker._sink_dynamo(self._generate_alerts(1))
...
```

Reply:
The client was already patched in the test setup, I was setting the mock return value to the patched object. However, I switched it to reference `boto_mock` instead of `self.forwarder.client_dynamo` to make it clearer (they are the same object).